
NAME: ADREETA GUHA
USC ID: 5122431956

In [68]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import precision_score

Q1

In [69]:
df=pd.read_csv("Caravan.csv")
df.head()

MOSTYPE  MAANTHUI  MGEMOMV  MGEMLEEF  MOSHOOFD  MGODRK  MGODPR  MGODOV  \
0       33         1        3         2         8       0       5       1   
1       37         1        2         2         8       1       4       1   
2       37         1        2         2         8       0       4       2   
3        9         1        3         3         3       2       3       2   
4       40         1        4         2        10       1       4       1   

   MGODGE  MRELGE  ...  APERSONG  AGEZONG  AWAOREG  ABRAND  AZEILPL  APLEZIER  \
0       3       7  ...         0        0        0       1        0         0   
1       4       6  ...         0        0        0       1        0         0   
2       4       3  ...         0        0        0       1        0         0   
3       4       5  ...         0        0        0       1        0         0   
4       4       7  ...         0        0        0       1        0         0   

   AFIETS  AINBOED  ABYSTAND  Purchase  
0       0        0         0        No  
1       0        0         0        No  
2       0        0         0        No  
3       0        0         0        No  
4       0        0         0        No  

[5 rows x 86 columns]

In [70]:
df.drop(["PVRAAUT","AVRAAUT"],axis=1,inplace=True)


In [71]:
df.columns.shape

(84,)

In [72]:
df.columns

Index(['MOSTYPE', 'MAANTHUI', 'MGEMOMV', 'MGEMLEEF', 'MOSHOOFD', 'MGODRK',
       'MGODPR', 'MGODOV', 'MGODGE', 'MRELGE', 'MRELSA', 'MRELOV', 'MFALLEEN',
       'MFGEKIND', 'MFWEKIND', 'MOPLHOOG', 'MOPLMIDD', 'MOPLLAAG', 'MBERHOOG',
       'MBERZELF', 'MBERBOER', 'MBERMIDD', 'MBERARBG', 'MBERARBO', 'MSKA',
       'MSKB1', 'MSKB2', 'MSKC', 'MSKD', 'MHHUUR', 'MHKOOP', 'MAUT1', 'MAUT2',
       'MAUT0', 'MZFONDS', 'MZPART', 'MINKM30', 'MINK3045', 'MINK4575',
       'MINK7512', 'MINK123M', 'MINKGEM', 'MKOOPKLA', 'PWAPART', 'PWABEDR',
       'PWALAND', 'PPERSAUT', 'PBESAUT', 'PMOTSCO', 'PAANHANG', 'PTRACTOR',
       'PWERKT', 'PBROM', 'PLEVEN', 'PPERSONG', 'PGEZONG', 'PWAOREG', 'PBRAND',
       'PZEILPL', 'PPLEZIER', 'PFIETS', 'PINBOED', 'PBYSTAND', 'AWAPART',
       'AWABEDR', 'AWALAND', 'APERSAUT', 'ABESAUT', 'AMOTSCO', 'AAANHANG',
       'ATRACTOR', 'AWERKT', 'ABROM', 'ALEVEN', 'APERSONG', 'AGEZONG',
       'AWAOREG', 'ABRAND', 'AZEILPL', 'APLEZIER', 'AFIETS', 'AINBOED',
       'ABYSTAND'

In [73]:
df_test=df[:1000]
df_test.shape

(1000, 84)

In [74]:
X_test=df_test.drop("Purchase",axis=1)

In [75]:
X_test.shape

(1000, 83)

In [76]:
X_test.columns

Index(['MOSTYPE', 'MAANTHUI', 'MGEMOMV', 'MGEMLEEF', 'MOSHOOFD', 'MGODRK',
       'MGODPR', 'MGODOV', 'MGODGE', 'MRELGE', 'MRELSA', 'MRELOV', 'MFALLEEN',
       'MFGEKIND', 'MFWEKIND', 'MOPLHOOG', 'MOPLMIDD', 'MOPLLAAG', 'MBERHOOG',
       'MBERZELF', 'MBERBOER', 'MBERMIDD', 'MBERARBG', 'MBERARBO', 'MSKA',
       'MSKB1', 'MSKB2', 'MSKC', 'MSKD', 'MHHUUR', 'MHKOOP', 'MAUT1', 'MAUT2',
       'MAUT0', 'MZFONDS', 'MZPART', 'MINKM30', 'MINK3045', 'MINK4575',
       'MINK7512', 'MINK123M', 'MINKGEM', 'MKOOPKLA', 'PWAPART', 'PWABEDR',
       'PWALAND', 'PPERSAUT', 'PBESAUT', 'PMOTSCO', 'PAANHANG', 'PTRACTOR',
       'PWERKT', 'PBROM', 'PLEVEN', 'PPERSONG', 'PGEZONG', 'PWAOREG', 'PBRAND',
       'PZEILPL', 'PPLEZIER', 'PFIETS', 'PINBOED', 'PBYSTAND', 'AWAPART',
       'AWABEDR', 'AWALAND', 'APERSAUT', 'ABESAUT', 'AMOTSCO', 'AAANHANG',
       'ATRACTOR', 'AWERKT', 'ABROM', 'ALEVEN', 'APERSONG', 'AGEZONG',
       'AWAOREG', 'ABRAND', 'AZEILPL', 'APLEZIER', 'AFIETS', 'AINBOED',
       'ABYSTAND'

In [77]:
Y_test=df_test.Purchase

In [78]:
Y_test.head()

0    No
1    No
2    No
3    No
4    No
Name: Purchase, dtype: object

In [79]:
df_train=df[1000:]
df_train.shape

(4822, 84)

In [80]:
X_train=df_train.drop("Purchase",axis=1)

In [81]:
X_train.shape

(4822, 83)

In [82]:
X_train.columns

Index(['MOSTYPE', 'MAANTHUI', 'MGEMOMV', 'MGEMLEEF', 'MOSHOOFD', 'MGODRK',
       'MGODPR', 'MGODOV', 'MGODGE', 'MRELGE', 'MRELSA', 'MRELOV', 'MFALLEEN',
       'MFGEKIND', 'MFWEKIND', 'MOPLHOOG', 'MOPLMIDD', 'MOPLLAAG', 'MBERHOOG',
       'MBERZELF', 'MBERBOER', 'MBERMIDD', 'MBERARBG', 'MBERARBO', 'MSKA',
       'MSKB1', 'MSKB2', 'MSKC', 'MSKD', 'MHHUUR', 'MHKOOP', 'MAUT1', 'MAUT2',
       'MAUT0', 'MZFONDS', 'MZPART', 'MINKM30', 'MINK3045', 'MINK4575',
       'MINK7512', 'MINK123M', 'MINKGEM', 'MKOOPKLA', 'PWAPART', 'PWABEDR',
       'PWALAND', 'PPERSAUT', 'PBESAUT', 'PMOTSCO', 'PAANHANG', 'PTRACTOR',
       'PWERKT', 'PBROM', 'PLEVEN', 'PPERSONG', 'PGEZONG', 'PWAOREG', 'PBRAND',
       'PZEILPL', 'PPLEZIER', 'PFIETS', 'PINBOED', 'PBYSTAND', 'AWAPART',
       'AWABEDR', 'AWALAND', 'APERSAUT', 'ABESAUT', 'AMOTSCO', 'AAANHANG',
       'ATRACTOR', 'AWERKT', 'ABROM', 'ALEVEN', 'APERSONG', 'AGEZONG',
       'AWAOREG', 'ABRAND', 'AZEILPL', 'APLEZIER', 'AFIETS', 'AINBOED',
       'ABYSTAND'

In [83]:
Y_train=df_train.Purchase

In [84]:
Y_train.head()

1000     No
1001    Yes
1002     No
1003     No
1004     No
Name: Purchase, dtype: object

a) A random forest model with 500 trees, max_features = 29, and random_state = 0.

In [85]:
RF=RandomForestClassifier(max_features = 29,
                                   n_estimators = 500,
                                   random_state=0)
RF.fit(X_train,Y_train)

RandomForestClassifier(max_features=29, n_estimators=500, random_state=0)

In [86]:
RF_Pred=pd.DataFrame(RF.predict(X_test),columns=['Purchase_Pred'])

In [87]:
RF_Pred.shape

(1000, 1)

In [88]:
pd.crosstab(RF_Pred.Purchase_Pred,Y_test)

Purchase        No  Yes
Purchase_Pred          
No             915   52
Yes             26    7

In [89]:
test_set_precision1=7/(26+7)
test_set_precision1*100

21.21212121212121

Test Set Precision for Random Forest is 21.21%.

b) A boosting model with max_depth = 4. Use 1000 trees, learning rate 0.01, and random_state = 0.

In [91]:
GB=GradientBoostingClassifier(n_estimators = 1000,
                                      learning_rate = 0.01,
                                      max_depth = 4, random_state =0)
GB.fit(X_train,Y_train)
GB_Pred=pd.DataFrame(GB.predict(X_test),columns=['Purchase_Pred'])
GB_Pred.shape

(1000, 1)

In [92]:
pd.crosstab(GB_Pred.Purchase_Pred,Y_test)

Purchase        No  Yes
Purchase_Pred          
No             937   57
Yes              4    2

In [93]:
test_set_precision2=2/(2+4)
test_set_precision2*100

33.33333333333333

Test Set Precision for Gradient Boosting is 33.33%.

c) A multilayer perceptron with three hidden layers having 400 nodes each as follows
MLPClassifier(solver=’lbfgs’,random_state = 0,
activation = ’tanh’,hidden_layer_sizes=[400,400,400])

In [94]:
scaler = StandardScaler()
scaler.fit(X_train);
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_test_scaled[:5,:5]

array([[ 0.68458477, -0.27590228,  0.40588175, -1.21715837,  0.78272613],
       [ 0.99671296, -0.27590228, -0.85925192, -1.21715837,  0.78272613],
       [ 0.99671296, -0.27590228, -0.85925192, -1.21715837,  0.78272613],
       [-1.18818438, -0.27590228,  0.40588175,  0.00659844, -0.97080732],
       [ 1.2308091 , -0.27590228,  1.67101541, -1.21715837,  1.48413952]])

In [95]:
MLP=MLPClassifier(solver='lbfgs',random_state = 0,
activation = 'tanh',hidden_layer_sizes=[400,400,400],max_iter = 10000)
MLP.fit(X_train_scaled,Y_train)
MLP_Pred=pd.DataFrame(MLP.predict(X_test_scaled),columns=['Purchase_Pred'])
MLP_Pred.shape

(1000, 1)

In [96]:
pd.crosstab(MLP_Pred.Purchase_Pred,Y_test)

Purchase        No  Yes
Purchase_Pred          
No             883   47
Yes             58   12

In [97]:
test_set_precision3=12/(12+58)
test_set_precision3*100

17.142857142857142

Test set precision for MLP is 17.14%.

d) Logistic regression with the default value for C.

In [98]:
LR=LogisticRegression(solver = 'lbfgs',max_iter=10000,random_state = 0)
LR.fit(X_train_scaled,Y_train)
LR_Pred=LR.predict(X_test_scaled)
LR_Pred.shape

(1000,)

In [99]:
pd.crosstab(LR_Pred,Y_test)

Purchase   No  Yes
row_0             
No        934   59
Yes         7    0

In [100]:
precision=0/(0+7)
precision

0.0

Precision for logistic regression is 0%.

e) Repeat part (d) with a cut-off probability value equal to 0.25. LogisticRegression uses 0.50 as
cut-off default value. It uses the cut-off value to predict the category. If the probability is larger than
0.5 the model predicts category Yes, otherwise it predicts No. Function model.predict_proba( )
may help to this end.

In [101]:
LR_Pred1=pd.DataFrame(np.where(LR.predict_proba(X_test_scaled)[:,1]>0.25,
                               'Yes','No'),columns=['Predictions'])
LR_Pred1

Predictions
0            No
1            No
2            No
3            No
4            No
..          ...
995          No
996          No
997          No
998          No
999          No

[1000 rows x 1 columns]

In [102]:
pd.crosstab(LR_Pred1.Predictions,Y_test)

Purchase      No  Yes
Predictions          
No           917   48
Yes           24   11

In [103]:
precision_score(Y_test,LR_Pred1.Predictions,pos_label='Yes')

0.3142857142857143

In [104]:
test_set_precision4=11/(11+24)
test_set_precision4*100

31.428571428571427

Test Set Precision is 31.43% approx.

Q2.

In [105]:
from sklearn.datasets import fetch_lfw_people
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
faces = fetch_lfw_people(min_faces_per_person = 60)


a) (10 pts.) Fit an SVC model as shown. Find the test accuracy rate.
svc = SVC(kernel=’rbf’, gamma = 0.01)
model = make_pipeline(svc)
model.fit(Xtrain, ytrain);

In [106]:

from sklearn.pipeline import make_pipeline

In [107]:
from sklearn.metrics import confusion_matrix

In [108]:
print(faces.images.shape,faces.data.shape,faces.target.shape,faces.target_names.shape)


(1348, 62, 47) (1348, 2914) (1348,) (8,)


In [109]:
X = faces.data
y = faces.target

In [110]:
Xtrain, Xtest,ytrain, ytest = train_test_split(X, y,
                                train_size=0.75,
                                stratify=y,random_state = 42)


In [111]:
svc = SVC(kernel='rbf', gamma = 0.01)
model = make_pipeline(svc)
model.fit(Xtrain, ytrain);


In [112]:
Y_pred= model.predict(Xtest)
cm1=pd.crosstab(ytest,Y_pred,rownames=['Y_Test'],colnames=['Y_Pred'])
cm1.index=[faces.target_names]
cm1.columns=[faces.target_names[3]]
cm1

George W Bush
Ariel Sharon                  19
Colin Powell                  59
Donald Rumsfeld               30
George W Bush                133
Gerhard Schroeder             27
Hugo Chavez                   18
Junichiro Koizumi             15
Tony Blair                    36

In [114]:
accuracy_score(ytest,Y_pred,normalize=False)

133

In [115]:
accuracy_score(ytest,Y_pred,normalize=True)

0.39465875370919884

Accuracy rate is 39.46%.

b) (10 pts.) To improve the model accuracy we will transform the features using Principal Components,
a topic we review in ISE 535. Use first 150 principal components as the new features of the model,
as shown below, then report the test accuracy rate.

In [116]:
from sklearn.decomposition import PCA
pca = PCA(n_components=150, svd_solver='randomized',whiten=True,random_state=42)
svc = SVC(kernel='rbf', gamma = 0.01)
model1 = make_pipeline(pca,svc)
model1.fit(Xtrain, ytrain);


In [117]:
Y_Pred1=model1.predict(Xtest)

In [118]:
accuracy_score(ytest,Y_Pred1,normalize=True)

0.6973293768545994

Accuracy rate is 69.73%.

In [119]:
cm2=pd.crosstab(ytest,Y_Pred1,rownames=['Y_Test'],colnames=['Y_Pred1'])
cm2.index=[faces.target_names]
cm2.columns=[faces.target_names]
cm2

Ariel Sharon Colin Powell Donald Rumsfeld George W Bush  \
Ariel Sharon                 6            0               0            13   
Colin Powell                 0           50               0             9   
Donald Rumsfeld              0            0               8            22   
George W Bush                0            1               0           132   
Gerhard Schroeder            0            0               0            15   
Hugo Chavez                  0            0               0            14   
Junichiro Koizumi            0            0               0            10   
Tony Blair                   0            0               0            18   

                  Gerhard Schroeder Hugo Chavez Junichiro Koizumi Tony Blair  
Ariel Sharon                      0           0                 0          0  
Colin Powell                      0           0                 0          0  
Donald Rumsfeld                   0           0                 0          0  
George W Bush                     0           0                 0          0  
Gerhard Schroeder                12           0                 0          0  
Hugo Chavez                       0           4                 0          0  
Junichiro Koizumi                 0           0                 5          0  
Tony Blair                        0           0                 0         18

c) (10 pts.) Add argument class_weight=’balanced’ to the SVC() line above, run the code again and
find the accuracy rate. Here we modify the loss function by weighting the loss of each row category
by its class weight.


In [120]:
from sklearn.decomposition import PCA
pca = PCA(n_components=150, svd_solver='randomized',whiten=True,random_state=42)
svc = SVC(kernel='rbf',class_weight='balanced', gamma = 0.01)
model2 = make_pipeline(pca,svc)
model2.fit(Xtrain, ytrain);

In [121]:
Y_Pred2=model2.predict(Xtest)

In [122]:
accuracy_score(ytest,Y_Pred2,normalize=True)

0.744807121661721

The test accuracy rate is 74.48%.

d) (10 pts.) Use GridSearchCV to search best values for C and gamma. Try the grid values shown below.
Report the test accuracy rate.


In [123]:
pca = PCA(n_components=150, svd_solver='randomized',whiten=True,random_state=42)
svc = SVC(kernel='rbf', class_weight='balanced')
model3 = make_pipeline(pca,svc)
param_grid = {'svc__C': [1, 5, 10],
'svc__gamma': [0.0001, 0.0005, 0.001, 0.005]}
grid = GridSearchCV(model3, param_grid,cv=5,iid = True)
grid.fit(Xtrain, ytrain)


C:\Users\ADEETA\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:847: FutureWarning: The parameter 'iid' is deprecated in 0.22 and will be removed in 0.24.
  warnings.warn(


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('pca',
                                        PCA(n_components=150, random_state=42,
                                            svd_solver='randomized',
                                            whiten=True)),
                                       ('svc', SVC(class_weight='balanced'))]),
             iid=True,
             param_grid={'svc__C': [1, 5, 10],
                         'svc__gamma': [0.0001, 0.0005, 0.001, 0.005]})

In [124]:
grid.best_estimator_

Pipeline(steps=[('pca',
                 PCA(n_components=150, random_state=42, svd_solver='randomized',
                     whiten=True)),
                ('svc', SVC(C=5, class_weight='balanced', gamma=0.005))])

In [125]:
pca = PCA(n_components=150, svd_solver='randomized',whiten=True,random_state=42)
svc = SVC(kernel='rbf', class_weight='balanced',C=5,gamma=0.005)
model4 = make_pipeline(pca,svc)
model4.fit(Xtrain, ytrain)

Pipeline(steps=[('pca',
                 PCA(n_components=150, random_state=42, svd_solver='randomized',
                     whiten=True)),
                ('svc', SVC(C=5, class_weight='balanced', gamma=0.005))])

In [126]:
Y_Pred3=model4.predict(Xtest)

In [127]:
accuracy_score(ytest,Y_Pred3,normalize=True)

0.8486646884272997

Accuracy is 84.86%.

e) (10 pts.) Report the confusion matrix (for test set) of the best model in a dataframe with rows
identifying the true name of the person and columns identifying the predicted name. What person
has the largest error rate (that is, the person that is most difficult to predict)?


In [128]:
conf=pd.DataFrame(confusion_matrix(ytest,Y_Pred3),
                  index=faces.target_names,
                  columns=faces.target_names)
conf 

Ariel Sharon  Colin Powell  Donald Rumsfeld  George W Bush  \
Ariel Sharon                 10             4                0              5   
Colin Powell                  1            53                0              4   
Donald Rumsfeld               0             2               23              4   
George W Bush                 0             5                0            128   
Gerhard Schroeder             0             0                0              2   
Hugo Chavez                   0             3                0              3   
Junichiro Koizumi             0             0                1              2   
Tony Blair                    0             0                1              6   

                   Gerhard Schroeder  Hugo Chavez  Junichiro Koizumi  \
Ariel Sharon                       0            0                  0   
Colin Powell                       0            0                  0   
Donald Rumsfeld                    0            1                  0   
George W Bush                      0            0                  0   
Gerhard Schroeder                 24            0                  0   
Hugo Chavez                        1           10                  0   
Junichiro Koizumi                  0            0                 10   
Tony Blair                         1            0                  0   

                   Tony Blair  
Ariel Sharon                0  
Colin Powell                1  
Donald Rumsfeld             0  
George W Bush               0  
Gerhard Schroeder           1  
Hugo Chavez                 1  
Junichiro Koizumi           2  
Tony Blair                 28

In [129]:
error_rates=[]
conf.sum(axis=1)

Ariel Sharon          19
Colin Powell          59
Donald Rumsfeld       30
George W Bush        133
Gerhard Schroeder     27
Hugo Chavez           18
Junichiro Koizumi     15
Tony Blair            36
dtype: int64

In [130]:
for i in range(len(conf.index)):
    error_rates.append(1-conf.iloc[i,i]/conf.sum(axis=1)[i])
error_rates

[0.4736842105263158,
 0.10169491525423724,
 0.23333333333333328,
 0.03759398496240607,
 0.11111111111111116,
 0.4444444444444444,
 0.33333333333333337,
 0.2222222222222222]

In [131]:
max(error_rates)

0.4736842105263158

In [132]:
error_rates.index(max(error_rates))

0

In [133]:
conf.index[0]

'Ariel Sharon'

Ariel Sharon has the largest error rate.